Paper: Codebraid: Live Code in Pandoc Markdown

gpoore opened the pull request:
I will be doing additional proofreading over the next few days.

I am also working on adding some additional features to the software. Depending on how that goes, I may try to add a few additional paragraphs later this week.

gpoore committed:
initial commit; first draft

gpoore committed:
added section on working with external files

gpoore committed:
rewrapped long line of code

gpoore committed:
added JavaScript to list of supported languages

gpoore committed:
added docutils reference

gpoore committed:
minor proofreading

gpoore committed:
proofread bib file, fixed some en dashes and capitalization, made things more uniform

stefanv commented:
:wave: Hi @gpoore, I will be reviewing your paper.

Before I read through the text in detail, I wonder if you've run across any of the following (I don't see them mentioned in the text):

- https://github.com/mwouts/jupytext
- https://github.com/aaren/notedown
- `jupyter nbconvert`
- https://github.com/matthew-brett/nb2plots
- Jupyter Book

Their authors may also be good reviewers for this paper: @mwouts @aaren @carreau @matthew-brett @choldgraf

gpoore commented:
@stefanv I believe I have run across all of those. I'd appreciate input on how much detail to add about markdown-notebook programs, since there are a lot of them (more listed below!) and I'm still debating the best way to cover them. Currently, the introduction is focusing on Jupyter notebooks, knitr, and Org Babel, with a few mentions of additional programs and extensions. I could add some brief mentions of additional markdown-notebook programs. An alternative would be to restructure things somewhat and include a more in-depth comparison. The paper is a little more than 2 pages under maximum length, so there's easily space to add another page or so of detailed comparison if that's useful. Either way, I could probably rework the introduction/comparisons within a day or so.

I need to add discussion of `nbconvert` and cite it explicitly. Just as I was making the last revisions, I learned about global content filtering and tag-based filtering of cells/output. So parts of the introduction aren't up-to-date on `nbconvert` capabilities.

There actually is a brief citation of Jupyter Book (`[LH19]`)...it might be good to mention Jupyter Book explicitly in the text, and this part will need to be revised in any case since it relates to `nbconvert`.

In terms of Jupytext, notedown, and nb2plots: I do need to add more about other markdown-notebook programs. In addition to those, I'm also aware of these (plus [Pweave](http://mpastell.com/pweave/) and [Weave.jl](http://weavejl.mpastell.com/stable/) that are already mentioned in the paper):
* https://github.com/rossant/ipymd
* https://github.com/podoc/podoc
* https://github.com/jankatins/knitpy
* https://github.com/pystitch/stitch
* https://github.com/jhrmnn/knitj
* https://code.visualstudio.com/docs/python/jupyter-support and https://atom.io/packages/hydrogen

Also, while I've been working on this project over the last several months, Pandoc has added ipynb as a supported format (https://pandoc.org/demos.html, example 15). While I currently mention that in the conclusion, I should also add some comparison with it earlier.

matthew-brett commented:
I'm happy to help with review.

stefanv commented:
Tip of the hat for mentioning org-mode.  And, yes, pandoc is an obvious one I left out.  I think a review sections that outlines the landscape of these tools would be very helpful for readers and potential users.  It would also give an even stronger motivation for your work.

I'll set myself a reminder to do another review here on Monday; does that seem like a good timeline?

@matthew-brett Thank you for agreeing to join in the review!

gpoore commented:
@stefanv That sounds good. I'll rework the introduction and add an in-depth review of these additional tools, and have that ready for Monday.

@matthew-brett Thanks for reviewing!

mwouts commented:
Hello, I am not sure I can do an in-depth review, but I am happy to give a few comments...
First, thanks @gpoore for working on this, and thanks @stefanv for reaching out. This is a very interesting project. 

Obviously I like the fact that it works with plain Markdown files! Also, I appreciate the possibility to write code over multiple cells. 

I do not know how many people actually mix languages in a single document, but I would say anyway that this is already possible with R and Python both in R Markdown (with `reticulate`) and in Python jupyter notebooks (with the `%%R` magic cells).

Regarding outputs, I would say that support for rich outputs (images and javascript) is very important, is this available in `codebraid` ?

Finally, @gpoore you do mention that you are interested in converting `codebraid` documents back and forth to Jupyter notebooks. Clearly this will is a great idea - please get in touch if you want that we work together on the subject.

gpoore committed:
updated code block syntax for latest codebraid version

gpoore commented:
@mwouts Thanks for the comments!

There isn't yet support for rich outputs, but that is planned. The lightweight, built-in code execution system can be extended to support some rich outputs. I expect to have Jupyter kernels working as an alternate execution system very soon, so that will automatically add a lot of rich output capabilities.

I don't know when I will get around to converting between Codebraid documents and Jupyter notebooks, but I would definitely be interested in working on that together when the time comes.

choldgraf commented:
hey all - what kind of timeline are we looking at for a review? I'm gonna be slammed until at least the end of next week, so wouldn't be able to look at anything before then.

matthew-brett commented:
I was going to say the same thing as @mwouts - that there is a only a small subset of literate-programming use-cases for scientific Python, that don't need plots, nice table displays and so on.

To add to this crowded market - more Sphinx tools:

* https://nbsphinx.readthedocs.io/en/latest/
* https://jupyter-sphinx.readthedocs.io/en/latest/

deniederhut commented:
@choldgraf initial reviews are due June 11, but the open review period extends for another couple of weeks. If @gpoore is okay with getting some comments after the deadline, then it's also fine by me.

In case it's helpful, you can see our review criteria listed here - https://github.com/scipy-conference/scipy_proceedings/blob/2019/review_criteria.md

gpoore commented:
@choldgraf @deniederhut I'm happy to receive additional comments after the initial reviews are due.

gpoore committed:
revised introduction and added review; proofreading

gpoore committed:
added section on new support for jupyter kernels, moved and reorganized implementation and debugging sections to account for this; added mention of jupyter kernels elsewhere as needed; added paragraph about importance of pandoc ast

gpoore committed:
added image for jupyter kernel example

gpoore committed:
edited conclusion

gpoore committed:
proofreading

gpoore commented:
I uploaded an updated paper over the weekend, and just added an additional proofread.

* The paper now begins with an in-depth review of additional tools that allow mixing text with executable code, with special focus on programs related to Markdown.

* I now have Jupyter kernels working as an alternative to the built-in code execution system. I added a short example of this. I also merged and slightly condensed the old code execution and debugging sections, and moved them earlier in the paper so that this works better with the kernel example. A new Codebraid release with kernel support will be on GitHub and then PyPI shortly.

stefanv commented:
Built paper at: http://procbuild.scipy.org/download/gpoore-2019

Carreau commented:
I'm going to see if I can review this before the deadline. 
It may be of interest to @MSeal and @mpacer.

mwcraig commented on papers/geoffrey_poore/geoffrey_poore.rst:
```diff
@@ -0,0 +1,1073 @@
+
+
+
+Review
```
This review is *excellent* 💯 but runs fairly long. I struggled a bit with what to suggest here because on the one hand, I liked the detail but, on the other hand, it defers any focus on codebraid until p. 4.  I also understand that the review section was added in response to earlier reviews and agree that a comparison with other tools is very helpful.

I suggest one or more of these possibilities:

1. Edit the review section to be shorter; I am happy to provide specific recommendations if you want to go this route. 
1. Add a "compare and contrast with codebraid" section to each of the package descriptions.
1. Add a table that compares features check-box style so that the reader gets a quick summary of the content of the review section.
1. Drastically shorten the review section, moving the excellent detailed descriptions you have to an appendix.

stefanv replied:
I agree with @mwcraig's assessment.  Sorry if I pushed you even *further* in this direction.  While I think it is important to measure the "competition" and how codebraid is different, there is too much here and it needs to be in a different place in the paper (we first want to hear about codebraid!).  A table with feature comparison is a great idea.  These details can then move out to an appendix.

gpoore replied:
@mwcraig Thanks for being part of the review! I'm finally back to where I can focus on this again. I'm going to focus on the introduction and review over the next day or two. I will have a few comments on some of your other notes today. I'm basically in agreement with all your suggested modifications on later parts of the paper.

@mwcraig @stefanv Thanks for all the suggestions about the introduction and review. I'm going to work toward having a basic Codebraid example much earlier in the paper. @stefanv made a good point below about the first example not just being far into the paper, but also getting bogged down with details about Pandoc.

The problem I keep having with review/comparisons is that it's complex to pin down exactly what is possible with other tools due to all the extensions. **What do you think about an introduction/review that includes a comparison table for major alternatives (say, Org Babel, knitr, Jupyter Notebook, Pweave) as well as a paragraph or two of comparison for each, focusing only on default features?** Then I can condense the rest of the current review material into an appendix that is more focused on extensions. If I make that only about what is possible with extensions (omit most info about how it's done), it can be much shorter.

I have a simple comparison table in the README (https://github.com/gpoore/codebraid). That is a nice format for dealing with many kinds of features, though it can get a little tricky for a few of the features that aren't binary (maybe I can add or rephrase some categories).

stefanv replied:
That sounds like a good plan to me, Geoffrey.

mwcraig replied:
Sounds good to me too @gpoore 😀 

I saw the table in the readme after I submitted the review, something like that would be nice if it fits, OK to just link to it if needed.

mwcraig commented on papers/geoffrey_poore/geoffrey_poore.rst:
```diff
@@ -0,0 +1,1073 @@
+   lightweight, low-overhead code execution system.  Alternatively, code can
+   be executed with Jupyter kernels.  Regardless of how code is executed, a
+   single document can involve multiple programming languages, as well as
+   multiple independent sessions or processes per language.  All code
```
The sentence that starts "All code extraction and output insertion...." could be removed from the abstract.

mwcraig commented on papers/geoffrey_poore/geoffrey_poore.rst:
```diff
@@ -0,0 +1,1073 @@
+.. https://cran.r-project.org/web/packages/JuliaCall/index.html
+.. https://rmarkdown.rstudio.com/lesson-2.html
+
+Sweave :cite:`Leisch2002` was one of the first programs for combining text
```
This should be revised to change "one of the first" to "an early" or something like that. I used [noweb](https://en.wikipedia.org/wiki/Noweb) in graduation school in the mid-late 90s and it was in turn based on [WEB](https://en.wikipedia.org/wiki/WEB) by Knuth. I wouldn't recommend adding information about noweb/WEB/orCWEB here, just remove or modify the "one of the first" bit.

gpoore replied:
There's an ambiguity in my wording here that I hadn't realized. I was thinking about code that is part of a document and is executed during the build process, rather than code that can be extracted from the document and potentially executed later. Was the code in noweb documents ever executed to produce data or figures that were then included in the document?

mwcraig replied:
I don't think it was, no. It was a way of generating two outputs (latex, source code) from one document.

mwcraig commented on papers/geoffrey_poore/geoffrey_poore.rst:
```diff
@@ -0,0 +1,1073 @@
+programs that combine Jupyter kernels with Markdown or other formats to
+provide a different or broader set of features than what is possible with a
+default notebook.  In some cases, advances in nbconvert capabilities have made
+the features provided by these programs less distinctive.
```
If this sentence stays in it would be helpful to point out which features have been incorporated into `nbconvert`.

mwcraig commented on papers/geoffrey_poore/geoffrey_poore.rst:
```diff
@@ -0,0 +1,1073 @@
+plots) are displayed by default, but there are options for hiding code or output, or reversing their order.  All code for a given Jupyter session can be converted into a script or a Jupyter notebook.
+
+A simple Codebraid example
+==========================
```
While I was re-reading the the "Review" section this morning it occurred to me that you have an advantage over most readers of this paper: you are intimately familiar with `codebraid` and its syntax for marking up documents, so when you provide snippets from other systems, you know immediately what the corresponding snippet would look like in `codebraid`.

It might be worth having, before that section, a snipper like the one below (or perhaps even a shorter one) marked up for `codebraid`.  Then you could include that same snippet as an example in each of the other systems you review. That would put the reader in a better position to understand the differences/similarities in markup between the systems.

mwcraig commented on papers/geoffrey_poore/geoffrey_poore.rst:
```diff
@@ -0,0 +1,1073 @@
+Creating examples
+=================
+
+The example in the last section was actually itself an example of using
```
Nice 😀

mwcraig commented on papers/geoffrey_poore/geoffrey_poore.rst:
````diff
@@ -0,0 +1,1073 @@
+   ```
+
+The default format of ``stdout`` is ``verbatim``, but this was specified
+just to be explicit. The other option is ``raw``, or interpreted as
````
What does `raw` do? Is it the same as `copied_markup`?

gpoore replied:
`raw` means that Pandoc interprets it as Markdown, rather than showing it verbatim.

mwcraig commented on papers/geoffrey_poore/geoffrey_poore.rst:
```diff
@@ -0,0 +1,1073 @@
+``hide`` takes the same options as ``show``, except that ``none`` is
+replaced by ``all`` and formats are not specified. Instead of overriding
+existing settings like ``show``, ``hide`` removes the specified display
+options from those that currently exist.
```
A brief example would be helpful here.

mwcraig commented on papers/geoffrey_poore/geoffrey_poore.rst:
```diff
@@ -0,0 +1,1073 @@
+
+Codebraid currently provides two options for executing code: an included
+lightweight, low-overhead code execution system which is used by
+default; and Jupyter kernels. Jupyter kernels are demonstrated in the
```
```suggestion
default, and Jupyter kernels. Jupyter kernels are demonstrated in the
```

mwcraig commented on papers/geoffrey_poore/geoffrey_poore.rst:
```diff
@@ -0,0 +1,1073 @@
+
+..
+
+Since line numbers in errors and warnings correspond to those in the
```
I don't intend that you implement this suggestion for this paper or before SciPy, but it would be helpful to also report the line number in the original Markdown source file and/or the name of the block that produced the error. I agree that having the line number within the block is also useful; I can imagine hunting through several similar blocks to find the one that produced the error, though.

gpoore replied:
I actually have that working already for errors like trying to use Codebraid options that don't exist. For errors produced by the code, I just have to figure out the best way to present the information. Ultimately, I also hope to use this mapping data to get SyncTeX working between Markdown source and PDF output, so you can do forward/inverse search just as if you had a LaTeX source.

mwcraig replied:
Nice -- would it makes sense to mention that future feature in the conclusion?

mwcraig commented on papers/geoffrey_poore/geoffrey_poore.rst:
````diff
@@ -0,0 +1,1073 @@
+
+.. code:: text
+
+   ```{.python .cb.nb jupyter_kernel=python3}
````
Does the `codebraid` command have to be `.cb.nb` when using a Jupyter kernel, or does `{.python .cb.run jupyter_kernel=python3}` also work?

gpoore replied:
At this point, I think every command except for `.cb.expr` works with Jupyter kernels (`.cb.expr` technically works, it just doesn't display anything). Ultimately, Jupyter kernels will work with all commands...I just need to sort out a little more of the logic so that it's all consistent.

mwcraig commented on papers/geoffrey_poore/geoffrey_poore.rst:
```diff
@@ -0,0 +1,1073 @@
+
+The built-in code execution system allows for any number of languages
+within a single document. This is also possible when Jupyter kernels are
+used instead.
```
I assume one uses the usual magics for this?

gpoore replied:
You can use magics with Jupyter kernels. But you can also use multiples kernels, or multiple sessions for one kernel type. I've made a note to clarify this.

mwcraig commented on papers/geoffrey_poore/geoffrey_poore.rst:
```diff
@@ -0,0 +1,1073 @@
+of code presentation.
+
+Incomplete units of code
+------------------------
```
Questions about this capability: 

1. Do the incomplete and complete blocks have to be consecutive code blocks? I understand one can have intervening Markdown text, the question is whether one could have an intervening code block.

2. Can the order of the complete and incomplete blocks be flipped so that the complete block comes before the incomplete one? Not entirely sure when this would come up, just curious. 
2.

gpoore replied:
1. One or more incomplete code blocks in a given session have to be followed by a complete code block in that session. But in between all of them, there can be any amount of Markdown or any number of code blocks that are executed in other sessions (not that that is necessarily a good idea).

2. The blocks do need to be in order. It's possible to make things more exciting by using `copy` or `include` to get code from elsewhere, but even then, the location where the code is actually inserted does have to yield incomplete followed by complete.

mwcraig replied:
Thanks -- I don't think this needs to go in the paper...like you said, it is an edge case at best.

mwcraig commented on papers/geoffrey_poore/geoffrey_poore.rst:
````diff
@@ -0,0 +1,1073 @@
+
+.. code:: text
+
+   ```{.python .cb.run session=long}
````
Can `session` names be any string?

gpoore replied:
Names are currently restricted to valid Python identifiers. I should probably mention that. One of the next capabilities is extracting all code in a session to a separate source file. Restricting session names will allow them to be used as default source names.

mwcraig commented on papers/geoffrey_poore/geoffrey_poore.rst:
````diff
@@ -0,0 +1,1073 @@
+
+   ```{.python .cb.run session=long}
+   import json
+   result = sum(range(100_000_000))
````
`100_000_000`: This is possible?! Mind blown 😍

mwcraig commented on papers/geoffrey_poore/geoffrey_poore.rst:
```diff
@@ -0,0 +1,1073 @@
+
+Sessions work with both Codebraid’s code execution system and Jupyter
+kernels; for example, it is possible to use multiple python3 kernels
+within a single document.
```
With Jupyter kernels do you do multiple sessions by using `session=` or by choosing a different kernel for each block?

gpoore replied:
I should add more detail here. Jupyter kernels use `session` to specify sessions. Specifying the kernel just determines which kernel is used to execute a given session. Multiple, completely independent sessions are possible for a given kernel.

mwcraig commented on papers/geoffrey_poore/geoffrey_poore.rst:
```diff
@@ -0,0 +1,1073 @@
+chunks with ``outside_main=true`` at the beginning of a session are used
+to overwrite the beginning of the ``main()`` template, while any chunks
+with ``outside_main=true`` at the end of the session are used to
+overwrite the end of the template. If all code chunks have
```
What is in the beginning and end of the template? Does this mean "the stuff before main" and "the stuff after main"?

gpoore replied:
That's correct. I will clarify that.

mwcraig commented on papers/geoffrey_poore/geoffrey_poore.rst:
````diff
@@ -0,0 +1,1073 @@
+   ```
+
+includes the original Markdown source for this paper, and then uses a
+regular expression to display only the first few lines of the current
````
Recommend:

```suggestion
regular expression to display only the first few lines of this
```
I originally read this to mean that the snippet would display the first few lines of the section in which it was embedded regardless of its title.

mwcraig commented on papers/geoffrey_poore/geoffrey_poore.rst:
```diff
@@ -0,0 +1,1073 @@
+
+Since the ``cb.code`` command is including content from elsewhere, it is
+used with an empty code block. Alternatively, a single empty line or a
+single line containing an underscore is allowed as a placeholder. This
```
Start a new paragraph with "This example included part..."

mwcraig commented on papers/geoffrey_poore/geoffrey_poore.rst:
```diff
@@ -0,0 +1,1073 @@
+example included part of a file using a single regular expression. There
+are also options for including everything starting with or starting
+after a literal string or regular expression, and for including
+everything before or through a literal string or regular expression.
```
A link to the relevant part of the documentation would be helpful here.

mwcraig commented on papers/geoffrey_poore/geoffrey_poore.rst:
```diff
@@ -0,0 +1,1073 @@
+
+
+Conclusion
+==========
```
Given the review at the beginning, a paragraph in the conclusion doing a brief compare/contrast with other tools would be helpful. This should emphasize the things that codebrain does (or does better) that the others don't do. That is, I'd recommend focusing it on what codebraid brings to the table, rather than on what the other systems do.

mwcraig commented on papers/geoffrey_poore/geoffrey_poore.rst:
```diff
@@ -0,0 +1,1073 @@
+with PythonTeX :cite:`Poore2015` to create a code execution system that could
+be used with multiple markup languages.  While Codebraid has focused thus far
+on Pandoc Markdown, little of it is actually Markdown-specific.  It should be
+possible to work with other markup languages supported by Pandoc, such as
```
Could one already convert to any of these other formats using pandoc? It seems like one  of the advantages of codebraid is that any output option that pandoc provides you get for free. I understand that here you mean that other *input* formats could be supported without much effort.

gpoore replied:
It should be possible to convert to any format supported by Pandoc. There are currently a few issues with converting Markdown to Markdown when Pandoc's raw blocks are involved, but I think that's straightforward to fix. The main advantage of other *input* formats is that you get different document syntax and different editor tools.

mwcraig reviewed (commented):
Hi  @gpoore 👋! I'm one of the reviewers assigned to this paper. I was exicted when I read the title because most of the code I wrote for my phd work in the mid- to late-90s was written using `noweb`, a literate programming package. 

I have a number of suggestions for the current text. The main one is to tighten up or perhaps rearrange the review section at the beginning. A more detailed comment on that is at the beginning of the review section below.

I don't think there is a link to the codebraid repo or documentation in here; adding one or the other would be helpful, perhaps in a footnote when you first mention codebraid in the body of the paper.

Looking forward to working with you and the other reviewers over the next couple weeks!

mwcraig commented:
I meant to add: if you would like to make codebraid available in conda-forge I can help with that. Not necessary by any means, but won't take long if you want to.

MSeal commented on papers/geoffrey_poore/geoffrey_poore.rst:
```diff
@@ -0,0 +1,1073 @@
+code execution system, such as the Org-mode integration mentioned above.
+
+A Jupyter Notebook can only have a single kernel, and thus only a single
+primary programming language in a single session or process.  This means that
```
As a side note, not strictly true. A notebook can have only a single kernel, but there are some experimental multi-language kernels (not open source yet afaik). But for all practical purposes it's true enough given open source options and practicality of intended use of kernels.

gpoore replied:
Thanks for your comments! The two exceptions I'm aware of are [Script of Scripts](https://vatlab.github.io/sos-docs/), which is a kernel that manages other kernels, and perhaps to some extent [BeakerX](http://beakerx.com/), or maybe the original Beaker. I very briefly mention these at one point.

MSeal commented on papers/geoffrey_poore/geoffrey_poore.rst:
```diff
@@ -0,0 +1,1073 @@
+execution. The template writes delimiters to stdout and stderr at the
+beginning of each code chunk. These delimiters are based on a hash of
+the code to avoid the potential for collisions. Once execution is
+complete, Codebraid parses stdout and stderr and uses these delimiters
```
I am guessing this assumes single execution thread and order preserved stdout/stderr? Not unreasonable, but was just curious.

gpoore replied:
Yes, those are essentially the assumptions. Technically, multiple threads are possible, as well as async or other things that wouldn't preserve stdout/stderr order, as long as these are completely confined to a single code block, or to a sequence of code blocks that are marked as incomplete that are followed by a complete block.

MSeal commented:
Glad I got cc'd to see this write-up. It gave me a much better idea about codebraid, and it appears to be a much closer parallel to R-Markdown than other options out there for users who appreciated that pattern.

mwouts commented on papers/geoffrey_poore/poore.bib:
```diff
@@ -0,0 +1,414 @@
+  title  = {Jupyter notebooks as {Markdown} documents, {Julia}, {Python} or {R} scripts},
+  year   = {2018--2019},
+  url    = {https://jupytext.readthedocs.io/},
+}
```
Jupytext is now a team work indeed (with 25 contributors already!). Yet I think in a bibliography I'd like to be mentioned among the authors. Or maybe, @gpoore , you could add an entry for the original announcement for Jupytext, at https://towardsdatascience.com/introducing-jupytext-9234fdff6c57 ?

gpoore replied:
That's a good point. I'll add you as an author, and probably add an entry for the original announcement as well.

stefanv commented on papers/geoffrey_poore/build_rst_from_md.sh:
```diff
@@ -0,0 +1,2 @@
+codebraid pandoc -f markdown -t rst --overwrite -o geoffrey_poore.rst poore.txt
```
Nice :D :+1: :rocket:

deniederhut replied:
Using the tool to talk about the tool 👏👏👏

stefanv commented on papers/geoffrey_poore/poore.txt:
```diff
@@ -0,0 +1,1021 @@
+
+   Codebraid executes designated inline code and code blocks in Pandoc
+   Markdown documents as part of the document build process.  It includes a
+   lightweight, low-overhead code execution system.  Alternatively, code can
```
I imagine code execution is not something I'd expect to have large overhead.  I will read further to see if there's anything special being done in this regard.

gpoore replied:
The point I was trying to make is that the system that is executing the code has little effect on the overall runtime. In the Jupyter kernel case, starting the kernel can have a noticeable impact on overall runtime if the code itself doesn't take long to run. I'll probably remove "low-overhead" here and then explicitly explain this in the revised introduction.

stefanv commented on papers/geoffrey_poore/poore.txt:
```diff
@@ -0,0 +1,1021 @@
+   combination of its original Markdown source, its code, the stdout or stderr
+   resulting from execution, or rich output in the case of Jupyter kernels.
+   There is also support for programmatically copying code or output to other
+   parts of a document.
```
It is self-evident for those of us who know Markdown, but it may be worth mentioning that output types include PDF, HTML, EPUB, etc.

stefanv commented on papers/geoffrey_poore/poore.txt:
```````diff
@@ -0,0 +1,1021 @@
+`copy=code1+code2`.  It would also be possible to run the code elsewhere:
+
+``````
+```{.cb.run copy=code1+code2}
```````
Ah, ok, looks like I should rather review the `.rst` file---heading there.

stefanv commented on papers/geoffrey_poore/geoffrey_poore.rst:
```diff
@@ -0,0 +1,1073 @@
+
+..
+
+Pandoc Markdown defines attributes for inline code and code blocks.
```
This is the *first introduction* to codebraid, but here we are digging into Pandoc semantics instead of getting a big picture overview of the project, its goals, features, etc.

stefanv reviewed (commented):
The paper currently heavily focuses on technical details, which I think is fine for the "body" of the paper.  But it would be good to think a bit more about the story that the paper should tell.  "Hey, here's my new document generator; I call it codebraid.  I wrote it because I wanted to do X.  There are other packages that do some parts of X, but codebraid is different in that it does A, B, and C.  Let me show you how some of these features work: here's an example of A, one of B, and one of C. Let's look at how codebraid compares to those other packages (more in Appendix A). Conclusion: That was codebraid, and here is why you should use it."

Finally, don't forget to add links to the source repo! (I might have missed them.)

gpoore committed:
converted review into shorter comparison plus appendix (includes Jupytext reference changes suggested by @mwouts); condensed abstract; new intro with examples

gpoore committed:
removed discussion of Pandoc AST from example section (duplicates knitr comparison); moved Pandoc attribute discussion from example section into its own section immediately afterward

gpoore committed:
modified section references due to new Pandoc attribute section

gpoore committed:
edited command and display option descriptions: added reference to new intro, proofreading

gpoore committed:
edited code execution section: made language consistent with new intro, proofreading

gpoore committed:
improved Jupyter kernel section: clarified multiple kernels in response to @mwcraig, split example into two code chunks to illustrate features better

gpoore committed:
advanced code execution section: proofreading

gpoore committed:
edited external file section: clarification in response to @mwcraig

gpoore committed:
addressed most remaining points from @mwcraig: clarified raw display, clarified multiple languages with Jupyter kernels, mentioned session name restrictions, clarified "template" with outside_main, split paragraph under including external files, revised conclusion to emphasize Codebraid features; removed unneeded cb.nb example under commands to fix page length

gpoore commented:
There is now a new introduction that provides examples at the very beginning. This is followed by just over a page of comparisons with knitr, Pweave, Org-mode Babel, and Jupyter Notebook, focusing on default features. A summary of extensions and other, similar programs is now in an appendix at the end (a little under 1 page). 

I split the discussion of Pandoc attribute syntax into its own section, so it doesn't get in the way of the example of building a simple document. Thanks to @stefanv for pointing out that issue. I also revised the conclusion to emphasize features in response to @mwcraig, and have addressed essentially all of @mwcraig's other comments.

This is currently right at 8 pages, which does not leave room for a comparison table. If the current page of comparison would benefit from the addition of a table, or if there are any points where an additional example would be particularly helpful, I can free up space by dropping the subsection on `outside_main` and/or condensing the section on the built-in code execution system.

choldgraf commented:
hey all - just a note that since @MSeal and @mwcraig and maybe @Carreau are reviewing this, I'll likely hold off on reviewing and will instead read through their own thoughts because this is a paper of interest to me! I'm particularly curious if / how functionality here could be woven in with Jupyter Book (or if the models these two projects follow is different enough that it doesn't make sense).

As folks mentioned RMarkdown - at some point it would be great if the Jupyter community could agree on a standard for a more high-powered flavor of markdown that we can start adopting across projects (maybe the answer is RMarkdown or pandoc markdown?). That's out of the scope of this project submission, but it's a point I think worth tackling as a community!

deniederhut commented:
Whoa this has been a great group of reviews! Thanks everyone 😄

stefanv commented on papers/geoffrey_poore/geoffrey_poore.rst:
```diff
@@ -0,0 +1,1053 @@
+   can be replaced by a display of any combination of its original Markdown
+   source, its code, the stdout or stderr resulting from execution, or rich
+   output in the case of Jupyter kernels.  There is also support for
+   programmatically copying code or output to other parts of a document.
```
This is a **very good** summary.

stefanv commented on papers/geoffrey_poore/geoffrey_poore.rst:
```diff
@@ -0,0 +1,1053 @@
+programmatically, so code can be executed at one location in a document and
+its output displayed elsewhere.
+
+Building a simple Codebraid document
```
I would move anything related to codebraid to before the comparison.

gpoore replied:
That does make things flow better. I've changed that, and added a paragraph at the end of the introduction that summarizes where the paper is going.

stefanv reviewed (commented):
The paper is shaping up very nicely, Geoffrey!  I made one structural suggestion still: to finish talking about codebraid and its features, before moving on to comparison.  But I think the edits massively improve the readability and "punch" of the paper.

gpoore committed:
moved comparison to end per suggestion by @stefanv; added transition paragraph at end of intro; fixed PyPI acronym in intro

gpoore committed:
proofreading

mwcraig commented:
I'll have a chance to take another look tonight -- looking forward to it based on all the comments!

mwcraig commented on papers/geoffrey_poore/poore.txt:
```diff
@@ -0,0 +1,1011 @@
+kernels ``:cite:`Kluyver2016` ``{=rst} to execute code.  The first code block
+that is executed with a given language can specify a kernel.  In the example
+below, the "``.cb.nb``" tells Codebraid to run the code and provide a
+"notebook" display that shows both code and output.
```
I think it would be very useful here to emphasize that there are two *separate* directives here. 

`.cb.nb` is essentially a shorthand for a particular set of display choices intended to reproduce the input/output pattern of a notebook. 

`jupyter_kernel=python3` opts in to running the code with a Jupyter kernel the primary benefit of which (in this case) is getting the rich output you would see in a notebook.

While the two are often used together (and I think are used together in all of the examples of `.cb.nb` in the text, they could be used separately.

I think you could address this by saying explicitly that the `jupyter_kernel` bit indicates running the code in the kernel and generating rich output. You already come quite close to saying this, but I think erring on the side of explicit over implicit would be helpful.

mwcraig commented on papers/geoffrey_poore/poore.txt:
````diff
@@ -0,0 +1,1011 @@
+```
+
+Using Codebraid to execute code as part of the document conversion process is
+as simple as replacing `pandoc` with `codebraid pandoc`:
````
There is a really unfortunately line break in the PDF between `codebraid` and `pandoc`. Forcing those two to be on the same line would be very nice.

mwcraig commented on papers/geoffrey_poore/poore.txt:
```diff
@@ -0,0 +1,1011 @@
+not be ideal from some perspectives, it is the cost of maintaining full
+compatibility with Pandoc Markdown syntax.
+
+Finally, the `cb.nb` command runs code in "notebook mode."  For inline code,
```
I think `runs` should change to `displays` here, unless I am misunderstanding what `.cb.nb` does. My understanding is that it displays both the input and the output, and could be used without a kernel as in:

```
```{.python .cb.nb}
print("Hello like a notebook!")
```

In a similar vein, I gather one could do something like this to run in a kernel but display only the rich output:

```
```{.python .cb.run jupyter_kernel=Python3}
%matplotlib inline
from matplotlib import pyplot as plt
plt.plot([0, 1, 4, 9, 16])
```

gpoore replied:
I've reworded this to involve both `runs` and `displays`. Having `displays` does help, and I want to keep `runs` to emphasize that the code is executed (since it isn't always, as with `cb.code`).

mwcraig commented on papers/geoffrey_poore/poore.txt:
```diff
@@ -0,0 +1,1011 @@
+These can be used to override the default display settings for each command.
+
+`show` takes any combination of the following options: `markup` (display
+Markdown source), `code`, `stdout`, `stderr`, and `none`.  There is also
```
I know you said this earlier, but could you mention again here what `code` does, like you did for `markup`?

mwcraig commented on papers/geoffrey_poore/poore.txt:
```diff
@@ -0,0 +1,1011 @@
+code.  Multiple options can be combined, such as `show=code+stdout+stderr`.
+Code chunks using `copy` can employ `copied_markup` to display the
+Markdown source of the copied code chunk.  When the `cb.expr` command is used,
+the expression output is available via `expr`.  `show` completely overwrites
```
Unfortunately, to the casual reader this bit

```
is available via `expr`. `show` completely overwrites
```

reads as "is available via `expr.show` completely overrides" which is confusing. How about changing the last sentence to  something like: 

Using `show` completely overwrites...

gpoore replied:
That's a very good point! I hadn't thought about this inadvertently resembling some sort of attribute syntax.

mwcraig commented on papers/geoffrey_poore/poore.txt:
```diff
@@ -0,0 +1,1011 @@
+the expression output is available via `expr`.  `show` completely overwrites
+the existing display settings.
+
+The display format can also be specified with `show`.  `stdout`, `stderr`, and
```
Similar to the previous comment, this read to me as "The display format can also be specified with `show.stdout`, `stderr`" and was coupled with a line break right after `stdout`. The combination took me several tries to parse correctly.

Suggestion:

```
The display format of `stdout`, `stderr`, and `expr` can also be specified with `show`. The formats are `raw` (interpreted as Markdown)...
```

mwcraig commented on papers/geoffrey_poore/poore.txt:
```diff
@@ -0,0 +1,1011 @@
+The Bash configuration file also specifies that the file extension `.sh`
+should be used, and provides another four lines of template code to enable
+`cb.expr`.  So far, the longest configuration file, for Rust, is less than
+fifty lines—counting empty lines.
```
Could you mention where in the repo an interested reader could find these configuration files if they wanted to see a full example? Note: It may be so obvious as to not need an explanation, didn't have a chance to take a look this morning.

gpoore replied:
The config files are under `languages/` in the repo. If you think that's worth mentioning, I'm happy to do that.

mwcraig replied:
Nope, that seems easy enough to find.

mwcraig commented on papers/geoffrey_poore/poore.txt:
```diff
@@ -0,0 +1,1011 @@
+Notice that `jupyter_kernel` was only needed (and only allowed) for the first
+code chunk.  The second code chunk is still using the same language
+(`.python`), so it shares the same kernel.  This Markdown results in
+displayed code plus a plot, just as it would within a Jupyter notebook:
```
What do you think about changing this example to be `.cb.run` to demonstrate that `.cb.nb` is not necessary? You could add a sentence or clause indicating that `cb.nb` would produce a notebook-like display if that were desired. It just occurred to me that one advantage of codebrain of a traditional notebook is the easy of displaying only the output but you have no example of that.

Alternatively, a sentence or clause indicating that this could be done with `.cb.run` in both places to display only the output would be helpful.

mwcraig reviewed (commented):
This is really excellent, @gpoore! It flows much more smoothly and makes a great case for the advantages of codebraid. There are a couple of minor readability issues and a suggestion to include an example in which `.cb.nb` is used without a kernel or in which `.cb.run` is used with a kernel to demonstrate that `.cb.nb` is really a shorthand for "run and display using a notebook-like format` while `jupyter_kernel` means "run in a kernel, producing rich output and display however the rest of the codebraid commands indiciate the output should be displayed."

I can see that `.cb.nb jupyter_kernel=python3` will be a common use case, but codebraid's ability to produce notebook-like display without a notebook or to include rich output while by default hiding the code is really a plus.

Once again, nice job on this!

gpoore committed:
addressed @mwcraig comments; removed reference to cb.nb example under Jupyter kernels that is now a cb.run example; added missing mention of formats for rich_output under discussion of show

mwcraig commented:
@deniederhut -- I think this paper is ready to go and should be included in the proceedings. 👏 🚀

stefanv commented:
@deniederhut I am also happy to sign off on the paper.

Thanks, @gpoore, for being so responsive; I really like where this landed!

mwcraig commented:
FYI @gpoore I just opened https://github.com/conda-forge/staged-recipes/pull/8615 to add codebraid and bespon to conda-forge. Please comment on that PR whether you are willing to be listed as a maintainer on the conda-forge packages.

gpoore commented:
@stefanv Thanks for being part of the review, especially for your very helpful comments on the overall structure of the paper!

@mwcraig Thanks for being part of the review, especially for all the detailed comments about things that could be ambiguous and examples that could be improved! 

I'll do a final detailed proofread in the next day or two to make sure I haven't missed any grammar, layout, or typography issues.

deniederhut commented:
Woohoo! Thanks everyone!
